Relax schematron on beams

kepper opened the pull request.

kepper committed:
relax schematron on beams

fixes #578. Relax schematron on beams with `@sameas` attribute, as done already with `@copyof`.
Co-authored by @riedde
Co-authored by @th-we
Co-authored by @pe-ro

rettinghaus reviewed (approved):
Makes sense

musicEnfanthen commented on source/modules/MEI.cmn.xml:
```diff
@@ -1346,10 +1346,10 @@
     <constraintSpec ident="When_not_copyof_beam_content" scheme="isoschematron">
       <constraint>
-        <sch:rule context="mei:beam[not(@copyof)]">
+        <sch:rule context="mei:beam[not(@copyof) and not(@sameas)]">
```
is it "AND" or "OR" ?

musicEnfanthen replied:
SInce the text says "or"

musicEnfanthen replied:
true AND true = true; 
true AND false = false; (rule does not apply when one attribute is not there but the other is)
true AND true = false;

ture OR true = true;
true OR false = true (rule applies when only one attribute is not there)
false OR false = false

Or do I mess up and and and and and and or?

ahankinson replied:
Wow. You win the gold star for that last sentence. Perfect. ⭐

pe-ro replied:
Good catch.  I think it's "or".

bwbohl replied:
No, it definitely has to be `and` because what we want to express in words is:

If a beam has neither `@copyof` nor `@sameas` it has to have at least two descendants.

With `or` the expression would evaluate true if just one of them was missing.

musicEnfanthen reviewed (changes requested):
~~Can be merged after fixing AND to OR~~ EDIT: See comments above

musicEnfanthen commented:
Fixes #578

rettinghaus commented on source/modules/MEI.cmn.xml:
```diff
@@ -1346,10 +1346,10 @@
     <constraintSpec ident="When_not_copyof_beam_content" scheme="isoschematron">
       <constraint>
-        <sch:rule context="mei:beam[not(@copyof)]">
+        <sch:rule context="mei:beam[not(@copyof) and not(@sameas)]">
```
```suggestion
        <sch:rule context="mei:beam[not(@copyof or @sameas)]">
```

rettinghaus replied:
I think this was meant.

bwbohl replied:
Please be aware that this XPath pattern is setting the context for the following `sch:assert`. This is to say only nodes matching this pattern will be tested against the XPath expression in `sch:assert`, being: 
```xpath 
count(descendant::*[local-name()='note' or local-name()='rest' or local-name()='chord' or local-name()='space']) &gt; 1
```
We only have to test for the presence of at least two child nodes if both attributes (`@copyof` and `@sameas`) are missing.

The following table illustrates the results of the two different XPath patterns in `sch:rule/@ocntext`.

|    | mei:beam[not(@copyof) _and_ not(@sameas)]  | mei:beam[not(@copyof) _or_ not(@sameas)]  |
|:--------|:-------|:-------|
| `<beam/>` |  match   | match   |
| `<beam copyof=""/>`   | no match   | match  |
| `<beam sameas=""/>`   | no match  | match   |
| `<beam copyof="" sameas=""/>`   | no match   |  no match   |

If this table holds `match` in a cell that means, the element will be tested, i.e. child `note`, `rest`, `chord`, and `space` will be counted. An error will be reported if this number is less than `2`.

The big difference in the XPath patterns is the two rows in the middle. While _AND_ would not count their child elements, _OR_ will, because one of the attributes is missing and thus the pattern matches those `beam`s.

rettinghaus replied:
Sorry. I looked only at the bit that was changed.

bwbohl replied:
Same here, didn't realise until now that you changed the parenthesis … Your proposed expression `mei:beam[not(@copyof or @sameas)]`delivers the same results as `mei:beam[not(@copyof) and not(@sameas)]`

musicEnfanthen commented on source/modules/MEI.cmn.xml:
```diff
@@ -1346,10 +1346,10 @@
           <sch:assert
             test="count(descendant::*[local-name()='note' or local-name()='rest' or               local-name()='chord' or local-name()='space']) &gt; 1"
-            >A beam without a copyof attribute must have at least 2 note, rest, chord, or space
+            >A beam without a copyof or sameas attribute must have at least 2 note, rest, chord, or space
```
```suggestion
            >A beam that contains neither a copyof nor sameas attribute must have at least 2 note, rest, chord, or space
```

bwbohl committed:
Apply suggestions from code review

With the more compact XPath expression `not(@copyof or @sameas)` and the Neither…nor in the description the intention now is much clearer.

Co-authored-by: Klaus Rettinghaus <klaus.rettinghaus@gmail.com>
Co-authored-by: Stefan Münnich <stefan.muennich@unibas.ch>
